cognite_toolkit/_cdf_tk/resource_classes/capabilities.py

```diff
@@ -0,0 +1,507 @@
+
+
+class AssetRootIDScope(Scope):
+    _scope_name = "assetRootIdScope"
+    root_ids: list[str]
+
+
+class ExperimentScope(Scope):
+    _scope_name = "experimentscope"
+    experiments: list[str]
+
+
+class SpaceIDScope(Scope):
+    _scope_name = "spaceIdScope"
+    space_ids: list[str]
+
+
+class PartitionScope(Scope):
+    _scope_name = "partition"
+    partition_ids: list[int]
+
+
+class LegacySpaceScope(Scope):
+    _scope_name = "spaceScope"
+    external_ids: list[str]
+
+
+class LegacyDataModelScope(Scope):
+    _scope_name = "dataModelScope"
+    external_ids: list[str]
+
+
+class Capability(BaseModel):
+    _capability_name: ClassVar[str]
+    scope: Scope
+
+    @field_validator("scope", mode="before")
+    @classmethod
+    def find_scope_cls(cls, data: Any) -> Scope:
+        annotation = cls.model_fields["scope"].annotation
+        if isinstance(annotation, UnionType):
+            valid_types = {s._scope_name for s in get_args(annotation)}
+        elif annotation is not None and issubclass(annotation, Scope):
+            valid_types = {annotation._scope_name}
+        else:
+            raise ValueError(f"Invalid scope annotation '{annotation}'")
+
+        name = next(iter(data.keys()))
+        if name not in valid_types:
+            raise ValueError(
+                f"invalid scope name '{name}'. Expected one of {humanize_collection(valid_types, bind_word='or')}"
+            )
+
+        return Scope.model_validate(data)
+
+    @model_validator(mode="wrap")
+    @classmethod
+    def find_capability_cls(cls, data: Any, handler: ModelWrapValidatorHandler[Self]) -> Self:
+        if cls is not Capability:
+            return handler(data)
+        if not isinstance(data, dict):
+            raise ValueError(f"Invalid capability data '{type(data)}' expected dict")
+        name, content = next(iter(data.items()))
+        if name not in _CAPABILITY_CLASS_BY_NAME:
+            raise ValueError(f"Invalid capability name '{name}'. Expected one of {_CAPABILITY_CLASS_BY_NAME.keys()}")
+        cls_ = _CAPABILITY_CLASS_BY_NAME[name]
+        return cast(Self, cls_.model_validate(content))
+
+    @model_serializer(mode="wrap", when_used="always", return_type=dict)
+    def include_capability_name(self, handler: SerializerFunctionWrapHandler) -> dict:
+        if self._capability_name is None:
+            raise ValueError("Capability name is not set")
+        serialized_data = handler(self)
+        return {self._capability_name: serialized_data}
+
+
+class AnalyticsAcl(Capability):
+    _capability_name = "analyticsAcl"
+    actions: list[Literal["READ", "EXECUTE", "LIST"]]
+    scope: AllScope
+
+
+class AnnotationsAcl(Capability):
+    _capability_name = "annotationsAcl"
+    actions: list[Literal["READ", "WRITE", "SUGGEST", "REVIEW"]]
+    scope: AllScope
+
+
+class AssetsAcl(Capability):
+    _capability_name = "assetsAcl"
+    actions: list[Literal["READ", "WRITE"]]
+    scope: AllScope | DataSetScope
```

Comment:
Are all of these classes manually created? Where is the source of truth for these? Could they be autogenerated from protos?

Reply:
These are manually created. Source of truth is the OPENAPI spec. They can unfortunately not be generated as Toolkit has some deviation, for example, all cases were the API expects an internal ID, Toolkit uses an external ID such that the same configuration can be used across multiple CDF projects.